File: agent/typescript/src/framework/core/agent.ts
/**
 * Core Agent Implementation
 * 
 * Provides the foundational Agent class for building
 * Claude-powered agents with tool use capabilities.
 */

import Anthropic from '@anthropic-ai/sdk';
import type { MessageParam, ToolResultBlockParam } from '@anthropic-ai/sdk/resources/messages';

/**
 * Tool definition for agent capabilities
 */
export interface Tool {
  /** Unique tool identifier */
  name: string;
  /** Human-readable description of what the tool does */
  description: string;
  /** JSON Schema defining the tool's input parameters */
  input_schema: {
    type: 'object';
    properties: Record<string, any>;
    required?: string[];
  };
  /** Async function that executes the tool */
  handler: (input: any) => Promise<any>;
}

/**
 * Configuration options for creating an Agent
 */
export interface AgentConfig {
  /** Display name for the agent */
  name: string;
  /** Claude model to use */
  model: string;
  /** Maximum tokens for responses */
  maxTokens: number;
  /** Array of tools available to the agent */
  tools: Tool[];
  /** System prompt defining agent behavior */
  systemPrompt: string;
  /** Optional: Maximum agentic loop iterations (default: 10) */
  maxIterations?: number;
}

/**
 * Result of an agent execution
 */
export interface AgentResult {
  /** Final text response from the agent */
  response: string;
  /** Number of agentic loop iterations */
  iterations: number;
  /** Tools that were invoked during execution */
  toolsUsed: string[];
  /** Total tokens used */
  totalTokens: {
    input: number;
    output: number;
  };
}

/**
 * Agent class implementing the agentic loop pattern
 */
export class Agent {
  private client: Anthropic;
  private config: AgentConfig;
  private conversationHistory: MessageParam[] = [];

  constructor(config: AgentConfig) {
    this.client = new Anthropic();
    this.config = {
      ...config,
      maxIterations: config.maxIterations ?? 10
    };
  }

  /**
   * Execute the agent with a user message
   */
  async run(userMessage: string): Promise<AgentResult> {
    const toolsUsed: string[] = [];
    let iterations = 0;
    let totalInputTokens = 0;
    let totalOutputTokens = 0;

    this.conversationHistory.push({
      role: 'user',
      content: userMessage
    });

    let response = await this.client.messages.create({
      model: this.config.model,
      max_tokens: this.config.maxTokens,
      system: this.config.systemPrompt,
      tools: this.config.tools.map(t => ({
        name: t.name,
        description: t.description,
        input_schema: t.input_schema
      })),
      messages: this.conversationHistory
    });

    totalInputTokens += response.usage.input_tokens;
    totalOutputTokens += response.usage.output_tokens;

    while (
      response.stop_reason === 'tool_use' &&
      iterations < this.config.maxIterations!
    ) {
      iterations++;

      const toolUseBlocks = response.content.filter(
        (block): block is Anthropic.ToolUseBlock => block.type === 'tool_use'
      );

      const toolResults: ToolResultBlockParam[] = await Promise.all(
        toolUseBlocks.map(async (toolUse) => {
          const tool = this.config.tools.find(t => t.name === toolUse.name);

          if (!tool) {
            return {
              type: 'tool_result' as const,
              tool_use_id: toolUse.id,
              content: `Error: Unknown tool "${toolUse.name}"`,
              is_error: true
            };
          }

          toolsUsed.push(tool.name);

          try {
            const result = await tool.handler(toolUse.input);
            return {
              type: 'tool_result' as const,
              tool_use_id: toolUse.id,
              content: typeof result === 'string'
                ? result
                : JSON.stringify(result, null, 2)
            };
          } catch (error) {
            const errorMessage = error instanceof Error ? error.message : String(error);
            return {
              type: 'tool_result' as const,
              tool_use_id: toolUse.id,
              content: `Error: ${errorMessage}`,
              is_error: true
            };
          }
        })
      );

      this.conversationHistory.push({
        role: 'assistant',
        content: response.content
      });
      this.conversationHistory.push({
        role: 'user',
        content: toolResults
      });

      response = await this.client.messages.create({
        model: this.config.model,
        max_tokens: this.config.maxTokens,
        system: this.config.systemPrompt,
        tools: this.config.tools.map(t => ({
          name: t.name,
          description: t.description,
          input_schema: t.input_schema
        })),
        messages: this.conversationHistory
      });

      totalInputTokens += response.usage.input_tokens;
      totalOutputTokens += response.usage.output_tokens;
    }

    const textBlock = response.content.find(
      (block): block is Anthropic.TextBlock => block.type === 'text'
    );
    const finalResponse = textBlock?.text ?? '';

    this.conversationHistory.push({
      role: 'assistant',
      content: response.content
    });

    return {
      response: finalResponse,
      iterations,
      toolsUsed: [...new Set(toolsUsed)],
      totalTokens: {
        input: totalInputTokens,
        output: totalOutputTokens
      }
    };
  }

  reset(): void {
    this.conversationHistory = [];
  }
}

File: INTeract-ive-Agent/python/src/framework/core/agent.py

import asyncio
from typing import List, Dict, Any, Optional, TypedDict, Callable, Awaitable
from anthropic import AsyncAnthropic
from anthropic.types import MessageParam

# Type definitions matching the TS implementation
class Tool(TypedDict):
    name: str
    description: str
    input_schema: Dict[str, Any]
    handler: Callable[[Any], Awaitable[Any]]

class AgentConfig(TypedDict):
    name: str
    model: str
    max_tokens: int
    tools: List[Tool]
    system_prompt: str
    max_iterations: Optional[int]

class AgentTokenUsage(TypedDict):
    input: int
    output: int

class AgentResult(TypedDict):
    response: str
    iterations: int
    tools_used: List[str]
    total_tokens: AgentTokenUsage

class Agent:
    """Core Agent implementation."""

    def __init__(self, config: AgentConfig):
        self.client = AsyncAnthropic()
        self.config = config
        self.config['max_iterations'] = config.get('max_iterations', 10)
        self.conversation_history: List[MessageParam] = []

    async def run(self, user_message: str) -> AgentResult:
        tools_used: List[str] = []
        iterations = 0
        total_input_tokens = 0
        total_output_tokens = 0

        self.conversation_history.append({
            "role": "user",
            "content": user_message
        })

        # Prepare tools for Anthropic API
        api_tools = [
            {
                "name": t['name'],
                "description": t['description'],
                "input_schema": t['input_schema']
            }
            for t in self.config['tools']
        ]

        response = await self.client.messages.create(
            model=self.config['model'],
            max_tokens=self.config['max_tokens'],
            system=self.config['system_prompt'],
            tools=api_tools,
            messages=self.conversation_history
        )

        # Accumulate tokens (approximate or extracted from usage)
        if response.usage:
            total_input_tokens += response.usage.input_tokens
            total_output_tokens += response.usage.output_tokens

        while response.stop_reason == 'tool_use' and iterations < self.config['max_iterations']:
            iterations += 1
            
            # Extract tool use blocks
            tool_use_blocks = [block for block in response.content if block.type == 'tool_use']
            
            tool_results = []
            
            for tool_use in tool_use_blocks:
                tool_def = next((t for t in self.config['tools'] if t['name'] == tool_use.name), None)
                
                if not tool_def:
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": tool_use.id,
                        "content": f"Error: Unknown tool {tool_use.name}",
                        "is_error": True
                    })
                    continue
                
                tools_used.append(tool_def['name'])
                
                try:
                    result = await tool_def['handler'](tool_use.input)
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": tool_use.id,
                        "content": result if isinstance(result, str) else str(result)
                    })
                except Exception as e:
                    tool_results.append({
                        "type": "tool_result",
                        "tool_use_id": tool_use.id,
                        "content": f"Error: {str(e)}",
                        "is_error": True
                    })

            self.conversation_history.append({
                "role": "assistant",
                "content": response.content
            })
            
            self.conversation_history.append({
                "role": "user",
                "content": tool_results
            })

            response = await self.client.messages.create(
                model=self.config['model'],
                max_tokens=self.config['max_tokens'],
                system=self.config['system_prompt'],
                tools=api_tools,
                messages=self.conversation_history
            )

            if response.usage:
                total_input_tokens += response.usage.input_tokens
                total_output_tokens += response.usage.output_tokens

        # Extract final response
        text_block = next((block for block in response.content if block.type == 'text'), None)
        final_response = text_block.text if text_block else ""

        self.conversation_history.append({
            "role": "assistant",
            "content": response.content
        })

        return {
            "response": final_response,
            "iterations": iterations,
            "tools_used": list(set(tools_used)),
            "total_tokens": {
                "input": total_input_tokens,
                "output": total_output_tokens
            }
        }
    
    def reset(self):
        self.conversation_history = []
